KEP-4420: Retry Generate Name (#4421)
kubernetes/enhancements: keps/sig-api-machinery/4420-retry-generate-name/README.md

on webhooks with this feature.

###### What steps should be taken if SLOs are not being met to determine the problem?

Look at the number of resources (using kube-state-metrics or similar). If there
exists a very large number of resources (millions) of a particular resource kind
and they all share the same generateName prefix, this might be triggering
high retry load.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

This code is in a critical system codepath, if anything is wrong with it, it could cause
severe problems, up to and including data corruption and loss.

## Alternatives

There are many alternatives that we considered including:

> - Increasing the number of characters appended to generateName when generating

Reviewer: I assume number of chars is the length, but what about the number of possible chars? is that an option?

Author: It's an alternative. But it is a breaking change in two ways:

- there is a 63 char name length limit. With the 5 char random suffix length, a 58 char length limit is set on generateName. We'd have to change that (or change the max length limit name!) to make room for a change in random suffix length.
- there are major downstream projects that use "{5}" regexes to match the random suffix.

Author: I've clarified in the KEP

  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
- [ ] (R) Graduation criteria is in place
  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

> This enhancement proposes that create requests using generateName are retried automatically by the apiserver when the generated name conflicts with an existing resource name.

Reviewer 2: ```suggestion
This enhancement proposes that create requests using `generateName` are retried automatically by the API server when the generated name conflicts with an existing resource name.
```

## Motivation

> Kubernetes [generateName](https://dev-k8sref-io.web.app/docs/common-definitions/objectmeta-/) [generates](https://github.com/jpbetz/kubernetes/blob/4b5ee802e847a3ea8243946aeae588906bd6f79f/staging/src/k8s.io/apiserver/pkg/storage/names/generate.go#L49) a 5 char random suffix that is appended to the generateName prefix.

Reviewer 2: ```suggestion
Kubernetes [generateName](https://dev-k8sref-io.web.app/docs/common-definitions/objectmeta-/) [generates](https://github.com/jpbetz/kubernetes/blob/4b5ee802e847a3ea8243946aeae588906bd6f79f/staging/src/k8s.io/apiserver/pkg/storage/names/generate.go#L49) a 5 char random suffix that is appended to a prefix
typically specified as `.metadata.generateName`
```

Each char in the sequence is from the ["bcdfghjklmnpqrstvwxz2456789" range](https://github.com/jpbetz/kubernetes/blob/4b5ee802e847a3ea8243946aeae588906bd6f79f/staging/src/k8s.io/apimachinery/pkg/util/rand/rand.go#L83C15-L83C42).

This leads to $`27^5 = 14,348,907`$ possible names per generateName prefix. For

> context, we have seen clusters store 100,000..1,000,000 CRDs, so the total

Reviewer 2: ```suggestion
context, we have seen clusters store 100,000..1,000,000 custom resources, so the total
```

number of names that can be generated is not currently a problem.

But there is a problem with name conflicts. There is a 50% chance of hitting a
conflict before generating even 5000 names, and a 0.1% chance at only 500
names!

Today a HTTP 409 respose returned to the client if a name conflict happens.  It is
possible for clients to detect the error and retry, but we've seen that most
clients don't realize to do this and hit this problem in production.

### Goals

- Reduce probability of generated name conflicts to below 0.1% for up to 1

>   million generated names per generateName prefix.

Reviewer 2: (nit)
```suggestion
  million generated names per `generateName` prefix.
```

### Non-Goals

> - Change the length of the random suffix. This is a breaking change:
>   - There is a 63 char name length limit. to guarantee a 5 char random suffix
>     length, generateName is trimmed if its length is >58 chars. If the random
>     suffix length were to be increased, generateName would need to trimmed to a
>     shorter length, breaking users that depend on 58 chars of generateName being
>     preserved. (Increasing the max lengh of name would also be a breaking
>     change).
>   - There are major downstream projects that use "{5}" regexes to match the
>     random suffix.
> - Eliminate the possibility of generated name conflicts entirely. This is
>   much harder problem given that generateName does create names in a reserved
>   identifier space, and so any mecanism to pick names is prone to conflicts.

Reviewer 2: ```suggestion
- Change the length of the random suffix. This is a breaking change:
  - There is a 63 char name length limit. to guarantee a 5 char random suffix
    length, `generateName` is trimmed if its length is >58 chars. If the random
    suffix length were to be increased, `generateName` would need to trimmed to a
    shorter length, breaking users that depend on 58 chars of `generateName` being
    preserved. (Increasing the max length of `name` would also be a breaking
    change).
  - There are major downstream projects that use regular expressions with a `{5}` exact match
    count that is intended to match the random suffix.
- Eliminate the possibility of generated name conflicts entirely. This is
  much harder problem given that `generateName` does create names in a reserved
  identifier space, and so any mechanism to pick names is prone to conflicts.
```

## Proposal

Modify the apiserver to retry name generation. When performing the create

> operation at the storage layer of the kube-apiserver, if the crewate fails

Reviewer 3: Nit:
```suggestion
operation at the storage layer of the kube-apiserver, if the create fails
```

with a "name conflict" error, generate a new name and retry.

> If we retry up to 7 times, we can generate up to 1 million names before reaching a 0.1% chance of a collision. This is roughtly the same probability of collision we would get if we were to increased the number of chars per random generateName suffix to 11.

Reviewer 3: ```suggestion
If we retry up to 7 times, we can generate up to 1 million names before reaching a 0.1% chance of a collision. This is roughly the same probability of collision we would get if we were to increased the number of chars per random generateName suffix to 11.
```

> While this doesn't eliminate the possibility of name conflicts when very large
> numbers of resources are generated with the same generateName prefix, it does
> effectively eliminate name conflicts for more typical use cases. When generating
> <=100,000 resources with a single generateName prefix, there is a 1 in
> $`1.6x10^{13}`$ chance of ever encountering a name conflict.

Reviewer 2: (nit)
```suggestion
While this doesn't eliminate the possibility of name conflicts when very large
numbers of resources are generated with the same `generateName` prefix, it does
effectively eliminate name conflicts for more typical use cases. When generating
<=100,000 resources with a single `generateName` prefix, there is a 1 in
$`1.6x10^{13}`$ chance of ever encountering a name conflict.
```

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: RetryGenerateName
  - Components depending on the feature gate: kube-apiserver

###### Does enabling the feature change any default behavior?

No. It only reduces the probability of a error that can occur randomly.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes.

###### What happens if we reenable the feature if it was previously rolled back?

This is safe. The feature becomes enabled again.

###### Are there any tests for feature enablement/disablement?

Yes.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

N/A

###### What specific metrics should inform a rollback?

`apiserver_request_duration_seconds` for create requests.  If this metric
increases significantly when this feature is enabled, it may indicate
an issue with this feature.

We also plan to add a `generate_name_retries: {count}` label to this metric
to make it easy to inspect the latency of create requests where retries

> occurred.

Reviewer 2: Would there be a way for someone to determine what proportion of `generateName` **create** requests needed at least one retry? (I don't think this metric would provide that).

Reviewer 4: > Would there be a way for someone to determine what proportion of `generateName` **create** requests needed at least one retry? (I don't think this metric would provide that).

What would we suggest a cluster-admin do with the information that varies from "if you see a lot of retries, the feature may be having a problem"?

Author: Almost certainly the feature has a problem. There is also a small possibility that millions of names were generated with the same prefix and collisions have becoming common.. this requires an exotic etcd setup or an alternate storage backend, but it is possible..  I'll clarify.

Should I simplify the metric to just track if retries happened, and not track the count?

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### How can an operator determine if the feature is in use by workloads?

Presence of the `generate_name_retries: {count}` label on the `apiserver_request_duration_seconds` metric
with a count greater than 1 would indicate this feature has been used.

###### How can someone using this feature know that it is working for their instance?

- [x] Other

>   - Details: The lack of name conflict errors when using generateName is the only user visible behavior, but it

Reviewer 2: ```suggestion
  - Details: The lack of name conflict errors when using `generateName` is the only user visible behavior, but it
```

###### What are the SLIs (Service Level Indicators) an operator can use to determine the health of the service?

<!--
Pick one more of these and delete the rest.
-->

- [x] Metrics
  - Metric name: `apiserver_request_duration_seconds` for creates

###### Are there any missing metrics that would be useful to have to improve observability of this feature?

No

### Dependencies

###### Does this feature depend on any specific services running in the cluster?

No

### Scalability

###### Will enabling / using this feature result in any new API calls?

Yes, if the resource being created with generateName has validating webhooks registered,
the validating webhooks will be called for each retry.

###### Will enabling / using this feature result in introducing new API types?

No

###### Will enabling / using this feature result in any new calls to the cloud provider?

No

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

This can happen.

> When a retry does occur, if the resource being created with generateName has
> validating webhooks registered, the validating webhooks will be called for each
> retry. For slow webhooks this could result in a significant increase to request
> latency for the create requests using generateName.

Reviewer 2: (nit)
```suggestion
When a retry does occur, if the resource being created with `generateName` has
validating webhooks registered, the validating webhooks will be called for each
retry. For slow webhooks this could result in a significant increase to request
latency for the create requests using `generateName`.
```

BTW same thing for in-process evaluation of ValidatingAdmissionPolicies.

---

We might want to make the retry count load adaptive. Maybe. One way to do that is to have the API server watch (as in local inotify) for a dynamic configuration snippet that controls this kind of setting, and then use a mechanism (ConfigMap? special systemd unit? whatever) to write an appropriate value to that dynamic overrides file.

For alpha, we absolutely don't need that mechanism.

Author: Yeah, I've been thinking about this a bit too.  We could either offer more controller, or we could optimize it further.  I'll capture all of this as potential future work (either as post alpha or post ga)

We expect retries to be very rare, even for large numbers of resources
(100k-1M), but if a cluster were somehow able to generate a very large number of

> resources for a generateName, say 2M or higher, then retries could become more

Reviewer 2: ```suggestion
resources for a `generateName`, say 2M or higher, then retries could become more
```

common. 

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

> No, at least not in all but the most pathologcial of situations.

Reviewer 2: (nit)
```suggestion
No, at least not in all but the most pathological of situations.
```

If a cluster was receiving a high load of create requests for generateName
resources with a large number (millions) of existing resources already generated

> with the same generateName prefix, this could cause up to a 7x increase in CPU.

Reviewer 2: (nit)
```suggestion
with the same `generateName` prefix, this could cause up to a 7x increase in CPU.
```

But we do not expect this to actually happen, primarily because etcd cannot
support these volumes of resources.

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

No

### Troubleshooting

###### How does this feature react if the API server and/or etcd is unavailable?

###### What are other known failure modes?

The only know failure modes are pathological. E.g. A runaway system creating
unbounded resources using generateName could eventually trigger amplified load
on webhooks with this feature.

###### What steps should be taken if SLOs are not being met to determine the problem?

Look at the number of resources (using kube-state-metrics or similar). If there
exists a very large number of resources (millions) of a particular resource kind

> and they all share the same generateName prefix, this might be triggering

Reviewer 2: ```suggestion
and they all share the same `generateName` prefix, this might be triggering
```

high retry load.

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

This code is in a critical system codepath, if anything is wrong with it, it could cause
severe problems, up to and including data corruption and loss.

## Alternatives

There are many alternatives that we considered including:

> - Increasing the length of the random suffix appended to generateName when generating

Reviewer 2: (nit)
```suggestion
- Increasing the length of the random suffix appended to `generateName` when generating
```

  a name. This would be a breaking change in two ways:

>   - There is a 63 char name length limit. to guarantee a 5 char random suffix
>     length, generateName is trimmed if its length is >58 chars. If the random
>     suffix length were to be increased, generateName would need to trimmed to a
>     shorter length, breaking users that depend on 58 chars of generateName being
>     preserved. (Increasing the max lengh of name would also be a breaking
>     change).

Reviewer 2: (nit)
```suggestion
  - There is a 63 char `name` length limit. To guarantee a 5 char random suffix
    length, `generateName` is trimmed if its length is >58 chars. If the random
    suffix length were to be increased, `generateName` would need to trimmed to a
    shorter length, breaking users that depend on 58 chars of `generateName` being
    preserved. (Increasing the max length of `name` would also be a breaking
    change).
```

  - There are major downstream projects that use "{5}" regexes to match the

>     random suffix.

Reviewer 2: The 5 character thing is not an API guarantee, so it's an unwelcome change but not breaking in that sense.

Author: > so it's an unwelcome change but not breaking in that sense

I agree.  This KEP started with the question:  "Can I stop the bleeding this problem is causing right now without breaking anyone?"  I still seems plausable to me that at some point we'd rip the bandaid off and make the length change, but it will take longer to agree on, implemented, and roll out than this.

  - Note that the selected design is statistically as good as increasing the
    characters appended from 5 to 11.
- After generating a name, check if the name exists in etcd and, if it does,
  generate another name.  See https://github.com/kubernetes/kubernetes/compare/master...jpbetz:kubernetes:retry-generate-name
  The downside to this approach is that it adds a round-trip to etcd for

>   all create requests that use generateName.

Reviewer 2: ```suggestion
  all create requests that use `generateName`.
```

- Add a "retry filter" to the apiserver's filter chain to retry the entire
  request. This would reinvoke the entire admission chain and increase
  the latency of request handling more significantly than the other approaches
  considered.

> - Introduce automatic retry logic into the client.

Reviewer 2: We could also do both; the client could signify its willingness to automatically retry (eg, with a special query string on the **create** request).

Author: I like it. I'll make a note.

- Consider further optimizing. One possible option:  If the
  first create attempt conflicts, do a loop to check etcd
  for all subsequent name generations to make sure the name
  doesn't exist before retrying the create request.  This
  would be more efficient for cases where multiple retries
  are required while still keeping the non-conflicting case
  fast.

#### GA

- Add `generate_name_retries: {count}` label to `apiserver_request_duration_seconds`
  metric.

### Upgrade / Downgrade Strategy

N/A. This only reduces the probability of name conflicts when a request is
served by an apiserver with this feature enabled.

### Version Skew Strategy

N/A. This only reduces the probability of name conflicts when a request is
served by an apiserver with this feature enabled.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: NameGenerationRetries
  - Components depending on the feature gate: kube-apiserver

###### Does enabling the feature change any default behavior?

> No. It only reduces the probability of a error that can occur randomly.

Reviewer 2: I think the answer might be “Yes”; the default is now to retry (and users can see this if they look at the relevant metrics).

-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html
-->

- Verify that retries are not attempted when feature is disabled
- Verify that up to max retries are attempted when conflicting names exist
- Verify that if max retries is exceeded that a already exists error is returned

##### e2e tests

This feature is not e2e testable. There is no predictable way to
trigger name collisions.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag
- Unit tests added. The `generic/registry/store_test.go` suite covers this functionality remarkably well.

#### Beta

- Gather feedback from developers and surveys
- Add integration tests
- Additional tests are in Testgrid and linked in KEP
- Benchmark worst case: Max retries are attempted for all create requests.
- Consider further optimizing. One possible option:  If the

>   first create attempt conflicts, do a loop to check etcd

Reviewer 4: this could be ok after the first one.  I would like the first attempt to be "free" though.
